### PR TITLE
Krm/tink server

### PR DESCRIPTION
**Pull request by micahhausler**
## Description

This PR adds a Kubernetes API backend to Tinkerell as a replacement for Postgres.

This PR builds on #563 and adds new API capabilities

## Checklist:

I have:

- [ ] updated the documentation and/or roadmap (if required)
- [ ] added unit or e2e tests
- [ ] provided instructions on how to upgrade

### CODE REVIEW

**Review by thebsdbox — DISMISSED** [review]
NIIIIIIIIIIIIIIIIICE, `/lgtm`

**Review by mmlb — APPROVED** [review]
skipped over virtual-worker stuff and everything else looks alright. Haven't really used bdd style tests and tooling but if that works for you then :+1:. `server` is looking pretty big, we should split it up so the k8s and sql backends are separate subpackages later I think.

**Comment by mmlb** [comment]
I've added do-not-merge just until #602 is merged.

File: cmd/tink-worker/worker/worker.go [hunk 1]
```diff
@@ -207,11 +207,11 @@ func (w *Worker) execute(ctx context.Context, wfID string, action *pb.WorkflowAc
 	if err != nil {
 		return st, errors.Wrap(err, "wait container")
 	}
-
-	l.With("status", st.String()).Info("container removed")
+	// Removing the below line, its duplicated from above
+	// l.With("status", st.String()).Info("container removed")
```
[the remark this answers is not in the document]
Comment (micahhausler) [reply cmd/tink-worker/worker/worker.go]: done

File: cmd/virtual-worker/worker/container_manager.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+func (m *fakeManager) StartContainer(_ context.Context, id string) error {
+	m.logger.With("containerID", id).Debug("starting container")
+	return nil
+}
+
+func (m *fakeManager) WaitForContainer(_ context.Context, id string) (pb.State, error) {
+	m.logger.With("containerID", id).Info("waiting for container")
+
+	r := rand.New(rand.NewSource(time.Now().UnixNano()))
```
[the remark this answers is not in the document]
Comment (micahhausler) [reply cmd/virtual-worker/worker/container_manager.go:55]: @sonatype-lift ignore

File: cmd/virtual-worker/worker/container_manager.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+	"time"
+
+	"github.com/packethost/pkg/log"
+	"github.com/tinkerbell/tink/cmd/tink-worker/worker"
+	pb "github.com/tinkerbell/tink/protos/workflow"
+)
+
+func getRandHexStr(length int) string {
+	// intentionally weak RNG. This is only for fake output
+	r := rand.New(rand.NewSource(time.Now().UnixNano()))
```
[the remark this answers is not in the document]
Comment (micahhausler) [reply cmd/virtual-worker/worker/container_manager.go:15]: @sonatype-lift ignore

File: cmd/virtual-worker/worker/container_manager.go [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+	r := rand.New(rand.NewSource(time.Now().UnixNano()))
+	time.Sleep(time.Duration(r.Intn(m.sleepJitterMs)+m.sleepMinimumMs) * time.Millisecond)
+
+	return pb.State_STATE_SUCCESS, nil
+}
+
+func (m *fakeManager) WaitForFailedContainer(_ context.Context, id string, failedActionStatus chan pb.State) {
+	m.logger.With("containerID", id).Info("waiting for container")
+	r := rand.New(rand.NewSource(time.Now().UnixNano()))
```
[the remark this answers is not in the document]
Comment (micahhausler) [reply cmd/virtual-worker/worker/container_manager.go:63]: @sonatype-lift ignore

File: cmd/virtual-worker/worker/container_manager.go [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+}
+
+func (m *fakeManager) RemoveContainer(_ context.Context, id string) error {
+	m.logger.With("containerID", id).Info("removing container")
+	return nil
+}
+
+func (m *fakeManager) PullImage(_ context.Context, image string) error {
+	m.logger.With("image", image).Info("pulling image")
+	r := rand.New(rand.NewSource(time.Now().UnixNano()))
```
[the remark this answers is not in the document]
Comment (micahhausler) [reply cmd/virtual-worker/worker/container_manager.go:75]: @sonatype-lift ignore

File: cmd/virtual-worker/worker/container_manager.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (micahhausler) [reply cmd/virtual-worker/worker/container_manager.go:15]: @sonatype-lift ignore

File: cmd/tink-server/main.go [hunk 6]
```diff
@@ -53,6 +56,9 @@ func (c *DaemonConfig) AddFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&c.CertDir, "cert-dir", "", "")
 	fs.StringVar(&c.HTTPAuthority, "http-authority", ":42114", "The address used to expose the HTTP server")
 	fs.BoolVar(&c.TLS, "tls", true, "Run in tls protected mode (disabling should only be done for development or if behind TLS terminating proxy)")
+	fs.BoolVar(&c.KubernetesResourceModel, "kubernetes-backend", false, "Feature flag to use the Kubernetes Resource Model")
+	fs.StringVar(&c.KubeconfigPath, "kubeconfig", "", "The path to the Kubeconfig. Only takes effect if `--kubernetes-backend=true`")
+	fs.StringVar(&c.KubeAPI, "kube-api", "", "The kubernetes API endpoitn. Only takes effect if `--kubernetes-backend=true`")
```
[the remark this answers is not in the document]
Comment (micahhausler) [reply cmd/tink-server/main.go:61]: Sure

File: server/kubernetes_unimplemented.go [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	pb "github.com/tinkerbell/tink/protos/workflow"
+)
+
+const (
+	ErrNotImplemented = "not implemented"
```
[the remark this answers is not in the document]
Comment (micahhausler) [reply server/kubernetes_unimplemented.go:11]: I'll make this a defined error

[the remark this answers is not in the document]
Comment (mmlb) [reply server/kubernetes_unimplemented.go:11]: Will that run up against the conversation started in https://github.com/tinkerbell/lint-install/pull/41?

[the remark this answers is not in the document]
Comment (micahhausler) [reply server/kubernetes_unimplemented.go:11]: I'll make it unexposed for now

File: server/kubernetes_unimplemented.go [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+func (s *KubernetesBackedServer) GetWorkflow(context.Context, *pb.GetRequest) (*pb.Workflow, error) {
+	return nil, errNotImplemented
+}
+
+// DeleteWorkflow will return a not implemented error.
+func (s *KubernetesBackedServer) DeleteWorkflow(context.Context, *pb.GetRequest) (*pb.Empty, error) {
+	return nil, errNotImplemented
+}
+
+// ListWOrkflows will return a not implemented error.
```
Comment (displague) [inline server/kubernetes_unimplemented.go:27]:
```suggestion
// ListWorkflows will return a not implemented error.
```

**Review by displague — APPROVED** [review]
I'm a little late on the review here. I've noted a few places we can iterate on.

I'm glad to see we're moving ahead 🚀

**Pull request merged**